src/preproc/mod.rs

```diff
@@ -18,12 +18,116 @@
  * along with this program. If not, see <http://www.gnu.org/licenses/>.
  */
 
+//! This module mimics the Wikidot preprocessor, which replaces certian character sequences to make
+//! them look better, or be easier to parse.
+
 mod typography;
 mod whitespace;
 
 #[cfg(test)]
 mod test;
 
+use regex::Regex;
+
+/// Helper struct to easily perform string replacements.
+#[derive(Debug)]
+pub enum Replacer {
+    /// Replaces any text matching the "repl" group,
+    /// (or the entire regular expression if "repl" is happening)
+    /// with the static string.
+    RegexReplace {
+        regex: Regex,
+        replacement: &'static str,
+    },
+
+    /// Takes text matching the regular expression, and replaces the exterior.
+    ///
+    /// The regular expression must return the content to be preserved in
+    /// capture group 1, and surrounds it with the `begin` and `end` strings.
+    ///
+    /// For instance, say:
+    /// * `regex` matched `[% (.+) %]`
+    /// * `begin` was `<(`
+    /// * `end` was `)>`
+    ///
+    /// Then input string `[% wikidork %]` would become `<(wikidork)>`.
+    RegexSurround {
+        regex: Regex,
+        begin: &'static str,
+        end: &'static str,
+    },
+}
+
+impl Replacer {
+    /// Replaces the text in the manner defined by its enum, using the buffer as a temporary space
+    /// to copy to.
+    fn replace(&self, text: &mut String, buffer: &mut String) {
+        use self::Replacer::*;
+
+        match *self {
+            RegexReplace {
+                ref regex,
+                replacement,
+            } => {
+                debug!(
+                    "Running regex regular expression replacement (pattern {}, replacement {})",
+                    regex.as_str(),
+                    replacement,
+                );
+
+                let mut offset = 0;
+
+                while let Some(capture) = regex.captures_at(text, offset) {
+                    let range = {
+                        let full_match = capture
+                            .get(0)
+                            .expect("Regular expression lacks a full match");
+                        let mtch = capture.name("repl").unwrap_or(full_match);
```

Comment:
If `full_match` is only used in the case of `capture.name("repl")` being `None`, we can avoid performing that operation by doing `.unwrap_or_else()`, which takes a closure.
```suggestion
                        let mtch = capture.name("repl").unwrap_or_else(|| capture.get(0).unwrap()); // alternative is full match
```